script/upload

```diff
@@ -0,0 +1,262 @@
+    jq -r '.upload_url' |
+    sed -e 's/{.*}//g'
+}
+
+# Find the release files for the given version.
+release_files () {
+  local version="$1"
+
+  [ -n "$version" ] || return 1
+
+  find bin/releases -name '*.tar.gz' -o -name '*386*.zip' -o \
+    -name '*amd64*.zip' -o -name '*.exe' -o -name 'sha256sums.asc' | \
+    grep -E "$version|sha256sums.asc" | \
+    grep -v "assets" | \
+    LC_ALL=C sort
+}
+
+# Format the body message and print the file which contains it to the standard
+# output.
+finalize_body_message () {
+  local version="$1"
+  local changelog="$2"
+
+  version=$(echo "$version" | sed -e 's/^v//')
+
+  # If you change the list of distributions here, change docker/run_dockers.bsh
+  # as well.
+  cat "$changelog" > "$WORKDIR/body-template"
+  cat <<EOM >> "$WORKDIR/body-template"
+## Packages
+
+Up to date packages are available on [PackageCloud](https://packagecloud.io/github/git-lfs) and [Homebrew](http://brew.sh/).
+
+[RPM RHEL 6/CentOS 6](https://packagecloud.io/github/git-lfs/packages/el/6/git-lfs-VERSION-1.el6.x86_64.rpm/download)
+[RPM RHEL 7/CentOS 7](https://packagecloud.io/github/git-lfs/packages/el/7/git-lfs-VERSION-1.el7.x86_64.rpm/download)
+[Debian 7](https://packagecloud.io/github/git-lfs/packages/debian/wheezy/git-lfs_VERSION_amd64.deb/download)
+[Debian 8](https://packagecloud.io/github/git-lfs/packages/debian/jessie/git-lfs_VERSION_amd64.deb/download)
+[Debian 9](https://packagecloud.io/github/git-lfs/packages/debian/stretch/git-lfs_VERSION_amd64.deb/download)
+
+## SHA-256 hashes:
+EOM
+
+  shasum -a256 $(release_files "$version") | \
+    ruby -pe '$_.chomp!' \
+    -e '$_.gsub!(/^([0-9a-f]+)\s+.*\/([^\/]+)$/, "**\\2**\n\\1\n\n")' | \
+    ruby -0777 -pe '$_.gsub!(/\n+\z/, "\n")' >> "$WORKDIR/body-template"
+
+  sed -e "s/VERSION/$version/g" < "$WORKDIR/body-template" > "$WORKDIR/body"
+  echo "$WORKDIR/body"
+}
+
+# Filter a list of files from standard input, removing entries found in the file
+# provided.
+filter_files () {
+  local filter="$1"
+
+  # If the filter file is empty (that is, no assets have been uploaded), grep
+  # will produce no output, and therefore nothing will be uploaded. That's not
+  # what we want, so handle this case specially.
+  if [ -s "$filter" ]
+  then
+    grep -vF -f "$filter"
+  else
+    cat
+  fi
+}
+
+# Upload assets from the release directory to GitHub. Only assets that are not
+# already existing should be uploaded.
+upload_assets () {
+  local version="$1"
+  local upload_url="$2"
+  local file desc base ct encdesc encbase
+
+  curl https://api.github.com/repos/$REPO/releases | \
+    jq -r '.[] | select(.name == "'"$version"'") | .assets | .[] | .name' \
+    > "$WORKDIR/existing-assets"
+
+  for file in $(release_files "$version" | filter_files "$WORKDIR/existing-assets")
+  do
+    base=$(basename "$file")
+    desc=$(categorize_asset "$base")
+    ct=$(content_type "$base")
+    encbase=$(ruby -ruri -e 'print URI.escape(ARGV[0])' "$base")
```

Comment:
Can we use `curl(1)`'s `--data-urlencode` here instead?

Reply:
No, I don't believe we can. `--data-urlencode` requires that we be passing those values as a body, not as a URL, as far as I understand. The body we're passing here is actually the file itself, so we need to have some way to encode the text.

Reply:
As a further note, this is one place where core Perl no longer provides a built-in mechanism for escaping since `CGI.pm` has been removed, so if we switched to Perl here, we'd need to write something like `encbase=$(perl -e 'print $ARGV[0] =~ s/([^A-Za-z0-9])/sprintf "%%%02x", $1/gre;')`.